firecracker/microvm.py

```diff
@@ -262,30 +267,68 @@ def enable_rootfs(self, path_on_host: str) -> str:
             rootfs_filename = Path(path_on_host).name
             jailer_path_on_host = f"/opt/{rootfs_filename}"
             os.link(path_on_host, f"{self.jailer_path}/{jailer_path_on_host}")
-            return jailer_path_on_host
+            return Path(jailer_path_on_host)
         else:
             return path_on_host
 
+    def mount_rootfs(self, path_on_host: Path) -> Path:
+        """Mount a rootfs to the VM.
+        """
+        self.mounted_fs = path_on_host
+        if not self.use_jailer:
+            return path_on_host
+
+        rootfs_filename = Path(path_on_host).name
+        jailer_path_on_host = Path(DEVICE_BASE_DIRECTORY) / rootfs_filename
+        if not (self.jailer_path / jailer_path_on_host).is_block_device():
+            device_vm_path = Path(DEVICE_BASE_DIRECTORY) / str(self.vm_id)
+            device_vm_path.mkdir(exist_ok=True, parents=True)
+            path_to_mount = device_vm_path / rootfs_filename
+            if not path_to_mount.is_symlink():
+                path_to_mount.symlink_to(path_on_host)
+            # Copy all the /dev/dm-* special block files to make the mapping work on Jailer
+            os.system(f"cp -vap /dev/dm* {self.jailer_path}/dev/")
+            jailer_device_path_mount = Path(f"{self.jailer_path}/dev/mapper")
+            if not jailer_device_path_mount.is_mount():
+                # Mount as bind the /dev/mapper folder because if we change it as hardlinks it doesn't work
+                subprocess.run(
+                    ["mount", "--bind", "/dev/mapper", jailer_device_path_mount],
+                    check=True,
+                    capture_output=True)
+
+            jailer_path_to_mount = Path(f"{self.jailer_path}/{DEVICE_BASE_DIRECTORY}")
+            if not jailer_path_to_mount.is_mount():
+                # Mount and rbind (recursively bind) /dev/aleph-mapper folder to the Jailer path, but only for the
+                # target VM devices
+                subprocess.run(
+                    ["mount", "--rbind", device_vm_path, jailer_path_to_mount],
+                    check=True,
+                    capture_output=True)
+                os.system(f"chown -Rh jailman:jailman {self.jailer_path}/dev")
+
+        return jailer_path_on_host
+
```

Comment:
We must ensure that only the device mappers relevant to a VM are available inside its jailer.